File: extractors/triple_schema.py
"""Schema for extracted triples."""

from typing import Literal
from pydantic import BaseModel, field_validator

VALID_PREDICATES = frozenset([
    "cita", "supera", "usa", "trata_conceito", "define", "contradiz", "estende"
])


class ExtractedTriple(BaseModel):
    subject: str
    predicate: Literal["cita", "supera", "usa", "trata_conceito", "define", "contradiz", "estende"]
    obj: str
    source_paper: str
    confidence: float

    @field_validator("confidence")
    @classmethod
    def confidence_in_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError(f"confidence must be in [0, 1], got {v}")
        return v

    @field_validator("subject", "obj")
    @classmethod
    def not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("subject and obj must not be empty")
        return v

File: src/extractors/tripleSchema.ts
import { z } from "zod";

export const VALID_PREDICATES = [
  "cita",
  "supera",
  "usa",
  "trata_conceito",
  "define",
  "contradiz",
  "estende",
] as const;

export const ExtractedTripleSchema = z.object({
  subject: z.string().min(1),
  predicate: z.enum(VALID_PREDICATES),
  object: z.string().min(1),
  confidence: z.number().min(0).max(1).default(0.8),
  source: z.string().optional(),
});

export type ExtractedTriple = z.infer<typeof ExtractedTripleSchema>;
